### PR TITLE
feat(TransferList): Add transfer list component

### LINKED ISSUE #634: Add listbox / transfer list component
We should add a [listbox](https://www.w3.org/TR/wai-aria-practices-1.1/examples/listbox/listbox-rearrangeable.html) / [transfer list](https://mui.com/material-ui/react-transfer-list/) like component.  I think we should be able to use the styling present in `menu.scss` for the list boxes & our buttons for the left / right buttons.

Requested by Vikram More / SYNCHRO.

### PR DESCRIPTION
**Pull request by FlyersPh9**
## Changes

- Introduce new `TransferList` component which is used to move one or more items between lists.
```jsx
    <TransferList>
      <TransferList.ListboxWrapper>
        <TransferList.ListboxLabel>Listbox 1</TransferList.ListboxLabel>
        <TransferList.Listbox>
          <TransferList.Item>Item 1</TransferList.Item>          
          <TransferList.Item>Item 2</TransferList.Item>          
          <TransferList.Item>Item 3</TransferList.Item>
        </TransferList.Listbox>
      </TransferList.ListboxWrapper>
      <TransferList.Toolbar>
        <IconButton />
        <IconButton />
        <IconButton />
        <IconButton />
      </TransferList.Toolbar>
      <TransferList.ListboxWrapper>
        <TransferList.ListboxLabel>Listbox 2</TransferList.ListboxLabel>
        <TransferList.Listbox>
          <TransferList.Item>Item 4</TransferList.Item>          
          <TransferList.Item>Item 5</TransferList.Item>          
          <TransferList.Item>Item 6</TransferList.Item>
        </TransferList.Listbox>
      </TransferList.ListboxWrapper>
    </TransferList>
  );
```
- Allows for use to place anything they want above / below each transfer list box (example has search & checkbox).
- Using `.iui-list` from #1106.

Closes #634.

## Testing

- New unit tests and HTML demo page
- Referenced [w3 listbox article](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/examples/listbox-rearrangeable/) for accessibility best practices.

## Docs

- New changesets, website documentation, and stories added

### CODE REVIEW

**Comment by FlyersPh9** [comment]
In [the W3 example](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/examples/listbox-rearrangeable/) when hitting the tab key the focus goes to the listbox, not the individual listbox items.  The user can then use the keyboard up / down keys to select an item.  I think this would be preferred over how it is functioning now.  For example: if the left listbox had 100+ items, it would be painful to have to tab through all 100 items before focus moves over to the right listbox.

I imagine to make that work I'll need to make these changes in this PR:
- Add `@include iui-focus` to `@mixin iui-menu-listbox`.
- Add `tabindex="0"` to `<ul class="iui-menu" role="listbox">`.
- Remove `tabindex="0"` from all `<li class="iui-menu-item" role="option">`

And the rest of the functionality would be added in JS / React.  Is that a correct assumption?

File: packages/itwinui-css/src/utils/input-container/input-container.scss [hunk 1]
```diff
@@ -35,7 +35,8 @@
 
   .iui-input,
   .iui-input-with-icon,
-  .iui-input-group {
+  .iui-input-group,
+  .iui-menu {
```
Comment (FlyersPh9) [inline packages/itwinui-css/src/utils/input-container/input-container.scss:39]: Maybe this should be `.iui-menu[role='listbox']` instead?

**Comment by FlyersPh9** [comment]
I've re-run tests on GitHub a couple of times and I keep getting 1 failure.  I'm not getting the failure when running tests locally.  Unsure how to get this passing.

**Comment by mayank99** [comment]
I re-ran the tests, and CSS is passing entirely. I think the failures in React are legitimate. They are happening in the Select component, which is likely affected by your changes.

**Comment by elephantcatdog** [comment]
The open combobox and select tests are failing because of the missing shadow/border at the bottom.
image
image

File: packages/itwinui-css/src/menu/classes.scss [hunk 2]
```diff
@@ -4,6 +4,10 @@
 
 .iui-menu {
   @include iui-list-menu;
+
+  .iui-listbox-wrapper & {
```
Comment (FlyersPh9) [inline packages/itwinui-css/src/menu/classes.scss:8]: Thank you @elephantcatdog for the tip as to what was failing.  This change fixed the failed test images:
```diff
- &:where([role='listbox'])
+ .iui-listbox-wrapper & {
```

But depending on how #843 pans out, this may not be needed.  @mayank99

**Comment by mayank99** [comment]
@FlyersPh9 I've created #1106, which has very minimal changes, but will allow you to use `iui-list`/`iui-list-item` classes. Take a look and then we can think on how much styling belongs in which component.

I also see some screenshots in #634 which I feel would need a bit more work.

File: packages/itwinui-css/backstop/tests/listbox.html [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  </head>
+  <body class="iui-root">
+    <theme-button></theme-button>
+    <h1>Listbox</h1>
+    <hr />
+
+    <section id="demo-default">
+
+      <div class="iui-listbox-wrapper">
+        <div class="iui-input-container iui-with-message">
```
Comment (mayank99) [inline packages/itwinui-css/backstop/tests/listbox.html:43]: _input_-container is meant for _inputs_. we do not expose it publicly (it's marked internal) and i dont think we should complicate this component even more by trying to force it into listbox.

File: packages/itwinui-css/src/listbox/listbox.scss [hunk 4]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+// See LICENSE.md in the project root for license terms and full copyright notice.
+
+@mixin iui-listbox-wrapper {
+  --iui-menu-listbox-height: calc((var(--iui-component-height) - 1px) * 4.5);
```
Comment (mayank99) [inline packages/itwinui-css/src/listbox/listbox.scss:5]: why not just set `height`?

File: packages/itwinui-css/src/listbox/listbox.scss [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+  display: flex;
+  gap: var(--iui-size-xs);
+
+  > :where(:first-child, :last-child) {
+    flex: 1;
```
Comment (mayank99) [inline packages/itwinui-css/src/listbox/listbox.scss:11]: `:first-child`/`:last-child` selectors are not very reliable. maybe `grid` will be better as it will let you bypass the need for these.

Comment (FlyersPh9) [reply packages/itwinui-css/src/listbox/listbox.scss:11]: Yeah, `grid` might be a better choice here.  Using `grid` now.

File: packages/itwinui-css/src/listbox/listbox.scss (hunk 4 above)
Comment (FlyersPh9) [reply packages/itwinui-css/src/listbox/listbox.scss]: Trying to refresh my memory on this one, it's been a while.

I think I set the height on the `class="iui-list" role="listbox"` to avoid inconsistent heights if the listboxes contained a different amount of items:
Screenshot 2023-04-06 at 4 43 29 PM

And if `flex: 1` was applied but there was other stuff surrounding it:
Screenshot 2023-04-06 at 4 44 28 PM

Pink background applied to `.iui-listbox-wrapper`.

Comment (mayank99) [reply packages/itwinui-css/src/listbox/listbox.scss]: Let me rephrase the question. Why not set `height: calc((var(--iui-component-height) - 1px) * 4.5)` directly on the listbox rather than set a CSS variable here first and use it in the listbox?

I don't see how having different number of items affects it.

**Comment by adamhock** [comment]
The React portion of the PR is updated and ready for review.

File: packages/itwinui-css/backstop/tests/listbox.html [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    <theme-button></theme-button>
+    <h1>Listbox</h1>
+    <hr />
+
+    <section id="demo-default">
+
+      <div class="iui-listbox-wrapper">
+
+        <div class="iui-listbox-area">
+          <div class="iui-listbox-label">Options</div>
```
Comment (adamhock) [inline packages/itwinui-css/backstop/tests/listbox.html:47]: I placed `iui-listbox-label` inside of `iui-listbox-area` instead of listing them all at the top. This allowed me to get rid of an empty `<div>` that was placed between the labels and used to correct the formatting. The test images have changed slightly as a result

File: apps/website/src/pages/docs/listbox.mdx [hunk 7]
```diff
@@ -0,0 +1,49 @@
+---
+title: Listbox
+description: TODO Write Listbox description
```
Comment (adamhock) [inline apps/website/src/pages/docs/listbox.mdx:3]: There is still some TODOs for the documentation of this component. I'm not exactly sure how to describe this component without repeating the `List` documentation

File: packages/itwinui-css/src/listbox/listbox.scss (hunk 4 above)
Comment (adamhock) [reply packages/itwinui-css/src/listbox/listbox.scss]: Height is now being set directly in listbox and not through the use of a variable

File: packages/itwinui-css/backstop/tests/listbox.html (hunk 3 above)
Comment (adamhock) [reply packages/itwinui-css/backstop/tests/listbox.html]: Search box has been removed

File: packages/itwinui-css/src/listbox/listbox.scss (hunk 4 above)
Comment (mayank99) [reply packages/itwinui-css/src/listbox/listbox.scss]: yes, thank you!

**Comment by mayank99** [comment]
Currently it is missing all important interactions, such the ability to select items and navigate through them using keyboard. I'd imagine we'd also need to maintain state of currently selected items and provide props for user to control them.

Also I think the component should be renamed. "Listbox" by itself is only a small part of this whole pattern, and `role=listbox` is also used in other components (such as ComboBox). A better name might be "TransferList" (stolen from [mui](https://mui.com/material-ui/react-transfer-list/) / [mantine](https://mantine.dev/core/transfer-list/)).

File: apps/storybook/src/Listbox.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    className: { control: { disable: true } },
+    style: { control: { disable: true } },
+  },
+  title: 'Core/Listbox',
+} as Meta<ListboxProps>;
+
+export const Basic: Story<ListboxProps> = (args) => {
+  return (
+    <Listbox {...args}>
+      <Listbox.Area>
```
[the remark this answers is not in the document]
Comment (elephantcatdog) [reply apps/storybook/src/Listbox.stories.tsx:32]: Why can't you remove it? It looks the same without lines 32 & 41 to me.

File: apps/storybook/src/Listbox.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        <IconButton styleType={'borderless'}>
+          <SvgChevronDown />
+        </IconButton>
+        <IconButton styleType={'borderless'}>
+          <SvgChevronLeft />
+        </IconButton>
+        <IconButton styleType={'borderless'}>
+          <SvgChevronRight />
+        </IconButton>
+      </Listbox.Toolbar>
```
Comment (elephantcatdog) [inline apps/storybook/src/Listbox.stories.tsx:93]: This toolbar is centered based on the `Listbox` instead of between the two `Listbox.Areas`.
image
I think it should be centered between the two areas if it's meant to move items between them.

Comment (elephantcatdog) [reply apps/storybook/src/Listbox.stories.tsx:93]: Oh, I see it's centered in the html page.

File: apps/website/src/pages/docs/listbox.mdx (hunk 7 above)
Comment (elephantcatdog) [reply apps/website/src/pages/docs/listbox.mdx:3]: Maybe use something like "The Listbox component is used to display a `ListItem`s within the confines of a box"? I would just sprinkle "in a box" around haha

File: packages/itwinui-react/src/core/index.ts [hunk 10]
```diff
@@ -139,6 +139,9 @@ export type {
   ListItemDescriptionProps,
 } from './List';
 
+export { Listbox } from './Listbox';
```
[the remark this answers is not in the document]
Comment (elephantcatdog) [reply packages/itwinui-react/src/core/index.ts:142]: I put "listbox" in google and it seemed split on "Listbox" vs "ListBox". There was also "List Box/list box" in there. Seems fine either way to me.

**Comment by adamhock** [comment]
> Currently it is missing all important interactions, such the ability to select items and navigate through them using keyboard. I'd imagine we'd also need to maintain state of currently selected items and provide props for user to control them.
> 
> Also I think the component should be renamed. "Listbox" by itself is only a small part of this whole pattern, and `role=listbox` is also used in other components (such as ComboBox). A better name might be "TransferList" (stolen from [mui](https://mui.com/material-ui/react-transfer-list/) / [mantine](https://mantine.dev/core/transfer-list/)).

Implemented functionality for selecting items and navigating through them using the keyboard. The unit test for focusing in `List.test.tsx` is failing :/

I will push rename changes shortly

**Comment by mayank99** [comment]
> Implemented functionality for selecting items and navigating through them using the keyboard. The unit test for focusing in `List.test.tsx` is failing :/

There should be little to no changes in List and ListItem components. Both of those are generic low-level components that can be used in many places, so we cannot handle specific TransferList behaviors in there.

**Comment by adamhock** [comment]
> > Implemented functionality for selecting items and navigating through them using the keyboard. The unit test for focusing in `List.test.tsx` is failing :/
> 
> There should be little to no changes in List and ListItem components. Both of those are generic low-level components that can be used in many places, so we cannot handle specific TransferList behaviors in there.

This has been changed

**Comment by adamhock** [comment]
> The unit test for focusing in `List.test.tsx` is failing :/
> 

This unit test has been fixed

**Comment by mayank99** [comment]
> > > Implemented functionality for selecting items and navigating through them using the keyboard. The unit test for focusing in `List.test.tsx` is failing :/
> > 
> > 
> > There should be little to no changes in List and ListItem components. Both of those are generic low-level components that can be used in many places, so we cannot handle specific TransferList behaviors in there.
> 
> This has been changed

I see selection and arrow keys work internally, but currently there is no way for the user to access it. You need to expose the selection state and provide an easy way to implement a fully functional transferlist, such that items can be moved back and forth between the left and right listboxes and the state can be maintained by the user.

File: packages/itwinui-react/src/core/Listbox/Listbox.tsx [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+export type ListboxProps = {
+  /**
+   * Content in the listbox.
+   */
+  children: React.ReactNode;
+} & Omit<CommonProps, 'title'>;
+
+/**
+ * The Listbox container allows content to appear elevated through the use of a drop shadow
```
[the remark this answers is not in the document]
Comment (adamhock) [reply packages/itwinui-react/src/core/Listbox/Listbox.tsx]: This has been changed

**Comment by adamhock** [comment]
> > > > Implemented functionality for selecting items and navigating through them using the keyboard. The unit test for focusing in `List.test.tsx` is failing :/
> > > 
> > > 
> > > There should be little to no changes in List and ListItem components. Both of those are generic low-level components that can be used in many places, so we cannot handle specific TransferList behaviors in there.
> > 
> > 
> > This has been changed
> 
> I see selection and arrow keys work internally, but currently there is no way for the user to access it. You need to expose the selection state and provide an easy way to implement a fully functional transferlist, such that items can be moved back and forth between the left and right listboxes and the state can be maintained by the user.

The selection state is exposed and the toolbar story is now fully functional

File: apps/storybook/src/TransferList.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+    { name: 'Item 3', active: false },
+    { name: 'Item 4', active: false },
+    { name: 'Item 5', active: false },
+    { name: 'Item 6', active: false },
+  ]);
+
+  return (
+    <TransferList {...args}>
+      <TransferList.Area>
+        <TransferList.List role={'listbox'}>
```
Comment (gretanausedaite) [inline apps/storybook/src/TransferList.stories.tsx:43]: Why are you adding role='listbox'? I think that is already added in TransferList?

File: apps/storybook/src/TransferList.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+                }}
+              >
+                {item.name}
+              </TransferList.ListItem>
+            );
+          })}
+        </TransferList.List>
+      </TransferList.Area>
+      <TransferList.Toolbar>
+        <IconButton
```
Comment (gretanausedaite) [inline apps/storybook/src/TransferList.stories.tsx:182]: I think these buttons should be TransferList subcomponents.

File: packages/itwinui-css/src/transfer-list/transfer-list.scss [hunk 14]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+// See LICENSE.md in the project root for license terms and full copyright notice.
+@use '../input-container/mixins';
+@use './mixins' as *;
+
+.iui-transfer-list-wrapper {
+  display: grid;
+  grid-template-columns: 1fr auto 1fr;
```
Comment (gretanausedaite) [inline packages/itwinui-css/src/transfer-list/transfer-list.scss:8]: CSS suggestion: Maybe redo grid to have more than one row? Right now it messes up vertical centering when TransferList does not have message undeneath it.

File: packages/itwinui-react/src/core/List/List.tsx [hunk 15]
```diff
@@ -12,20 +12,26 @@ import type {
 import '@itwin/itwinui-css/css/menu.css';
 
 export const List = React.forwardRef((props, ref) => {
-  const { as: Element = 'ul', className, ...rest } = props;
+  const { as: Element = 'ul', className, role = 'list', ...rest } = props;
```
Comment (mayank99) [inline packages/itwinui-react/src/core/List/List.tsx:15]: undo changes in this file. `role` prop should automatically be respected as it will be part of `...rest`.

File: packages/itwinui-react/src/core/List/ListItem.tsx [hunk 16]
```diff
@@ -33,6 +33,7 @@ const ListItemComponent = React.forwardRef((props, ref) => {
       data-iui-size={size === 'large' ? 'large' : undefined}
       data-iui-actionable={actionable ? 'true' : undefined}
       data-iui-focused={focused ? 'true' : undefined}
+      tabIndex={disabled ? undefined : -1}
```
Comment (mayank99) [inline packages/itwinui-react/src/core/List/ListItem.tsx:36]: undo changes in this file. `tabIndex` should be managed by the consumer of this component.

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+
+export type TransferListAreaProps<T extends React.ElementType = 'div'> =
+  PolymorphicComponentProps<T, TransferListAreaOwnProps>;
+
+const TransferListArea = React.forwardRef((props, ref) => {
+  const { as: Element = 'div', children, className, ...rest } = props;
+
+  return (
+    <Element
+      className={cx('iui-transfer-list-area', className)}
```
Comment (mayank99) [inline packages/itwinui-react/src/core/TransferList/TransferList.tsx:34]: is it possible to remove this extra wrapper div?

ideally, the component would look like this:

```
<TransferList>
  <TransferList.Listbox>
   <TransferList.Item>...</TransferList.Item>
   <TransferList.Item>...</TransferList.Item>
   <TransferList.Item>...</TransferList.Item>
  </TransferList.Listbox>

  <TransferList.Toolbar>...</TransferList.Toolbar>

  <TransferList.Listbox>
   <TransferList.Item>...</TransferList.Item>
   <TransferList.Item>...</TransferList.Item>
  </TransferList.Listbox>
</TransferList>
```

File: apps/storybook/src/TransferList.stories.tsx [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+export default {
+  component: TransferList,
+  argTypes: {
+    className: { control: { disable: true } },
+    style: { control: { disable: true } },
+  },
+  title: 'Core/TransferList',
+} as Meta<TransferListProps>;
+
+export const Basic: Story<TransferListProps> = (args) => {
```
Comment (mayank99) [inline apps/storybook/src/TransferList.stories.tsx:30]: Does it make sense to show examples of transferlist with only one listbox? I'd imagine this component is pretty much always used with two listboxes.

File: apps/storybook/src/TransferList.stories.tsx (hunk 12 above)
Comment (adamhock) [reply apps/storybook/src/TransferList.stories.tsx]: This has been fixed

File: packages/itwinui-css/src/transfer-list/transfer-list.scss (hunk 14 above)
Comment (adamhock) [reply packages/itwinui-css/src/transfer-list/transfer-list.scss:8]: Genius. This has been added and the toolbar has been centered properly as a result

File: packages/itwinui-react/src/core/List/List.tsx (hunk 15 above)
Comment (adamhock) [reply packages/itwinui-react/src/core/List/List.tsx]: Fixed

File: packages/itwinui-react/src/core/List/ListItem.tsx (hunk 16 above)
Comment (adamhock) [reply packages/itwinui-react/src/core/List/ListItem.tsx]: `tabIndex` has been moved from `ListItem` to `TransferList.ListItem`, but I think it's still necessary to define it for the keyboard navigation. Correct me if I am wrong on that

Comment (mayank99) [reply packages/itwinui-react/src/core/List/ListItem.tsx]: yes, `TransferList.ListItem` is the "consumer" of this component, so that's the right place to set tabindex

Comment (mayank99) [reply packages/itwinui-react/src/core/List/ListItem.tsx]: I think we will want to use `-1` for all items though, including disabled items.

Disabled items need to be focusable and should have `aria-disabled`, but clicking or pressing <kbd>Enter</kbd> while focusing them should do nothing.

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+ *       <TransferList.ListItem>Item 3</TransferList.ListItem>
+ *       <TransferList.ListItem>Item 4</TransferList.ListItem>
+ *       <TransferList.ListItem>Item 5</TransferList.ListItem>
+ *       <TransferList.ListItem>Item 6</TransferList.ListItem>
+ *     </TransferList.List>
+ *   </TransferList.Area>
+ * </TransferList>
+ */
+export const TransferList = Object.assign(
+  React.forwardRef(
```
Comment (mayank99) [inline packages/itwinui-react/src/core/TransferList/TransferList.tsx:236]: move this component to a local variable, preferably near the top

also make it polymorphic

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+    }
+  };
+
+  return (
+    <ListItem
+      ref={ref}
+      onClick={onClickEvents}
+      onKeyDown={onKeyDown}
+      active={active}
+      actionable={actionable}
```
Comment (mayank99) [inline packages/itwinui-react/src/core/TransferList/TransferList.tsx:158]: i think this should `actionable` by default

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+
+export type TransferListListItemProps<T extends React.ElementType = 'li'> =
+  PolymorphicComponentProps<T, TransferListListItemOwnProps>;
+
+const TransferListListItem = React.forwardRef((props, ref) => {
+  const { actionable, disabled, onActiveChange, children, active, ...rest } =
+    props;
+
+  const onClickEvents = () =>
+    actionable && onActiveChange && onActiveChange(active ?? false);
```
Comment (mayank99) [inline packages/itwinui-react/src/core/TransferList/TransferList.tsx:135]: why is `onActiveChange` called with the stale state? I would expect it to be called with the next expected state

File: packages/itwinui-css/src/menu/mixins.scss [hunk 22]
```diff
@@ -187,3 +187,10 @@ $iui-active-outline: 1px solid var(--iui-color-border-accent);
     fill: var(--_iui-list-item-icon-fill, var(--iui-color-icon));
   }
 }
+
+@mixin iui-transfer-list {
```
Comment (mayank99) [inline packages/itwinui-css/src/menu/mixins.scss:191]: lets move this to `transfer-list.scss` and create a new class, e.g. `iui-transfer-list-listbox`

File: packages/itwinui-css/src/transfer-list/mixins.scss [hunk 23]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+// See LICENSE.md in the project root for license terms and full copyright notice.
+@mixin iui-transfer-list($area: null) {
```
Comment (mayank99) [inline packages/itwinui-css/src/transfer-list/mixins.scss:3]: Can we move all this code into `transfer-list.scss`? No need to create mixins when we can just put the code directly under the classes

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx (hunk 17 above)
Comment (mayank99) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx:34]: two things came to mind:
1. css grid could be used to keep the layout and get rid of the extra div
2. right now, it looks like the label is not even associated with the listbox (missing `aria-labelledby`), so maybe the `label` should be a prop on `TransferList.Listbox`.

File: packages/itwinui-css/src/transfer-list/transfer-list.scss (hunk 14 above)
Comment (gretanausedaite) [reply packages/itwinui-css/src/transfer-list/transfer-list.scss:8]: Is it valid case for users to add message under the second column? Is it possible to do that easily with current API?
image

File: packages/itwinui-react/src/core/List/ListItem.tsx (hunk 16 above)
Comment (adamhock) [reply packages/itwinui-react/src/core/List/ListItem.tsx]: This has been fixed

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx (hunk 19 above)
Comment (adamhock) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx]: Fixed

File: packages/itwinui-css/src/menu/mixins.scss (hunk 22 above)
Comment (adamhock) [reply packages/itwinui-css/src/menu/mixins.scss]: Fixed

File: packages/itwinui-css/src/transfer-list/mixins.scss (hunk 23 above)
Comment (adamhock) [reply packages/itwinui-css/src/transfer-list/mixins.scss]: Fixed

File: packages/itwinui-css/src/menu/mixins.scss (hunk 22 above)
Comment (mayank99) [reply packages/itwinui-css/src/menu/mixins.scss]: `role=listbox` needs to be brought back. It is needed, just not for styling.

File: packages/itwinui-css/src/transfer-list/transfer-list.scss (hunk 14 above)
Comment (adamhock) [reply packages/itwinui-css/src/transfer-list/transfer-list.scss:8]: This is easily possible to do with our API as long as there is an empty `div` element to take the place of the middle column

image
image

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx [hunk 24]
```diff
@@ -156,7 +162,9 @@ const TransferListListItem = React.forwardRef((props, ref) => {
       onKeyDown={onKeyDown}
       active={active}
       actionable={actionable}
-      tabIndex={disabled ? undefined : -1}
+      tabIndex={-1}
+      aria-disabled={disabled}
```
Comment (mayank99) [inline packages/itwinui-react/src/core/TransferList/TransferList.tsx:177]: Similar to `role=listbox` on the list, each of the listitems should have `role=option`. Also should have `aria-selected` when it's active.

See APG: https://www.w3.org/WAI/ARIA/apg/patterns/listbox/examples/listbox-rearrangeable/

After adding those, please also test in NVDA to confirm it's working.

Comment (adamhock) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx:179]: This has been added, but unit tests were not passing, so I still need to add those

File: packages/itwinui-css/src/menu/mixins.scss (hunk 22 above)
Comment (adamhock) [reply packages/itwinui-css/src/menu/mixins.scss]: Fixed

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx (hunk 17 above)
Comment (adamhock) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx:51]: 1. I tried using `grid-template-area` and I don't think we can get the desired layout since the grid involves more than one element with the same class name. The grid would have to be constructed like so:

`'label . label'`
`'listbox toolbar listbox'`

Correct me if I'm wrong on this since I'm not too familiar with `display: grid`

2. If `TransferList.Label` is replaced with a `label` prop, that would lead to the label not being polymorphic. Is that an issue? Can we add a second prop to change the label's element type?

Comment (mayank99) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx:51]: if the css grid is too complicated, then lets keep the extra divs, but we still need to associate the label using `aria-labelledby`. context seems like the right solution for this.
```jsx
// this will set id in context
<TransferList.Area>

  // this will read id from context, and set `id` attribute
  <TransferList.Label label='Blah' />

  // this will read id from context, and set `aria-labelledby`
  <TransferList.List>...</TransferList.List>
</TransferList.Area>
```

also renaming suggestions:
- TransferList.Area -> TransferList.ListboxWrapper
- TransferList.List -> TransferList.Listbox
- TransferList.Label -> TransferList.ListboxLabel
- TransferList.ListItem -> TransferList.Item

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx (hunk 24 above)
Comment (adamhock) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx]: The unit test has been added

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx (hunk 17 above)
Comment (adamhock) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx]: Context is now used to pass a `labelId` to the label and listbox subcomponents

These components have also been renamed

File: apps/storybook/src/Listbox.stories.tsx (hunk 9 above)
Comment (adamhock) [reply apps/storybook/src/Listbox.stories.tsx]: This centering issue has been fixed

**Comment by adamhock** [comment]
I will be making changes shortly that address the use of `polymorphic` and `useGlobals`

**Comment by mayank99** [comment]
Reposting https://github.com/iTwin/iTwinUI/pull/1073#discussion_r1203745795 for visibility:

> Both of ListboxLabels gets the same labelId and ListBoxes are not described correctly. Example I'm using:
> image
> 
> "Avalable" listbox
> image
> 
> "Selected" listbox
> image

**Comment by adamhock** [comment]
> Reposting [#1073 (comment)](https://github.com/iTwin/iTwinUI/pull/1073#discussion_r1203745795) for visibility:
> 
> > Both of ListboxLabels gets the same labelId and ListBoxes are not described correctly. Example I'm using:
> > image
> > "Avalable" listbox
> > image
> > "Selected" listbox
> > image

This has been fixed

File: .changeset/ninety-keys-double.md [hunk 25]
```diff
@@ -0,0 +1,18 @@
+---
+'@itwin/itwinui-css': minor
+'@itwin/itwinui-react': minor
+---
+
+Added `TransferList` component.
+
+```jsx
```
[the remark this answers is not in the document]
Comment (adamhock) [reply .changeset/ninety-keys-double.md]: A description has been added. Let me know if the description is too wordy

**Comment by adamhock** [comment]
> It seems that I can't reach second ListBox with keyboard (Tab).

I have fixed this by adding `tabIndex={0}` to `TransferList.Listbox`

File: apps/storybook/src/TransferList.stories.tsx [hunk 26]
```diff
@@ -0,0 +1,288 @@
...
+              >
+                {item.name}
+              </TransferList.Item>
+            );
+          })}
+        </TransferList.Listbox>
+      </TransferList.ListboxWrapper>
+      <TransferList.Toolbar>
+        <IconButton
+          styleType={'borderless'}
```
Comment (gretanausedaite) [inline apps/storybook/src/TransferList.stories.tsx:229]: nit: It would be nice if you added label to tool buttons. It would set an example for users.

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx [hunk 27]
```diff
@@ -0,0 +1,282 @@
...
+   * 	TransferList listbox label subcomponent
+   */
+  ListboxLabel: TransferListListboxLabel,
+  /**
+   * 	TransferList toolbar subcomponent
+   */
+  Toolbar: TransferListToolbar,
+});
+
+export const TransferListContext = React.createContext<
```
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/TransferList/TransferList.tsx:268]: Is it a valid use case for ListBox to have disabled state (not individual list items, but full box disabled)?

File: .changeset/ninety-keys-double.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (gretanausedaite) [reply .changeset/ninety-keys-double.md]: A bit wordy. A simple comma separated list of subcomponents should be enough.

File: apps/storybook/src/TransferList.stories.tsx (hunk 26 above)
Comment (adamhock) [reply apps/storybook/src/TransferList.stories.tsx:231]: Labels have been added for each of the buttons

File: .changeset/ninety-keys-double.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (adamhock) [reply .changeset/ninety-keys-double.md]: Kept the `TransferList` description and then listed the new subcomponents

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx (hunk 27 above)
Comment (mayank99) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx:268]: Not that I can think of, but we can always revisit later if someone requests it.

File: packages/itwinui-react/src/core/TransferList/TransferList.tsx (hunk 24 above)
Comment (adamhock) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx]: TransferListNVDA

Here is the NVDA test. Is it supposed to tell me when an item is selected?

Comment (mayank99) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx]: yes. I think you're missing `aria-multiselectable` (check the APG link above).

here's voiceover on safari with `aria-multiselectable`:

https://github.com/iTwin/iTwinUI/assets/9084735/d138346f-3594-4a4a-8be4-49612de67dcc

Comment (adamhock) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx]: `aria-multiselectable` has been added. I still need to apply `aria-activedescendant` to the listbox though

TransferListNVDAWithMultiSelect

Comment (mayank99) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx]: `aria-activedescendant` is useful when the focus is on a different element, e.g. in the case of ComboBox, the focus is on input rather than the options. In this case, the focus is actualy on the options, so it should be ok.

You can compare by testing that demo from APG using NVDA as well. The results should be very similar.

Comment (adamhock) [reply packages/itwinui-react/src/core/TransferList/TransferList.tsx]: TransferListNVDAW3Example

The results are very similar

**Review by mayank99 — APPROVED** [review]
LGTM
